_posts/2020-04-28-mocking.adoc

```diff
@@ -0,0 +1,375 @@
+    	this.invoiceNotificationService = invoiceNotificationService;
+    }
+
+    public Invoice generateSendInvoice(Long orderId) {
+        final Invoice invoice = invoiceService.generateForOrder(orderId);
+        if (invoice.isAlreadySent()) {
+            invoiceNotificationService.sendInvoice(invoice);
+        } else {
+        	invoiceNotificationService.notifyInvoiceAlreadySent(invoice);
+        }
+        return invoice;
+    }
+}
+----
+
+When testing the `generateSendInvoice` method we most likely don't want to use the actual `InvoiceNotificationService` as it would
+result in sending real notifications. With the old Quarkus approach one could "override"
+the `InvoiceNotificationService` in tests by adding the following bean in the test sources:
+
+[source,java]
+----
+@Mock
+public class MockInvoiceNotificationService implements InvoiceNotificationService {
+
+    public void sendInvoice(Invoice invoice) {
+
+    }
+
+    public void notifyInvoiceAlreadySent(Invoice invoice) {
+
+    }
+}
+----
+
+When Quarkus scanned this code, the use of `@Mock` would result in `MockInvoiceNotificationService` being used as
+the implementation of `InvoiceNotificationService` in every place where a `InvoiceNotificationService` bean was injected
+(in CDI terms this is called an injection point).
+
+Although this mechanism is fairly straightforward to use, it nonetheless suffers from a few problems:
+
+* A new class (or a new CDI producer method) needs to be used for each bean type that requires a mock. In a large application where a lot of
+mocks are needed, the amount of boilerplate code increases unacceptably.
+* There is no way for a mock to be used for certain tests only. This is due to the fact that beans that are annotated with `@Mock` are normal
+CDI beans (and are therefore used throughout the application).
+Depending on what needs to be tested, this can be very problematic.
+* There is a no out of the box integration with https://site.mockito.org/[Mockito], which is the de-facto standard for
+mocking in Java applications. Users can certainly use mockito (most commonly by using a CDI producer method), but there
+is boilerplate code involved.
+
+== New approach
+
+Starting with Quarkus `1.4`, users have the ability to create and inject per-test mocks for normal scoped CDI beans using
+`io.quarkus.test.junit.QuarkusMock`. Moreover, Quarkus provides out of the box integration with Mockito
+allowing for zero effort mocking of CDI beans using the `io.quarkus.test.junit.mockito.@InjectMock` annotation.
+
+=== Using QuarkusMock
+
+`QuarkusMock` provides the foundation for mocking normal scoped CDI beans and is also used under the hood by `@InjectMock`, so let us examine
+it first. The best way to do this is using an example:
+
+[source,java]
+----
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mockito;
+
+import io.quarkus.it.mock.MockableBean1;
```

Comment:
I think that we could skip these imports...

Reply:
OK, I'll remove them